models/wiki.go

```diff
@@ -242,3 +242,77 @@ func (repo *Repository) DeleteWikiPage(doer *User, wikiName string) (err error)
 
 	return nil
 }
+
+// UploadWikiFiles uploads files to wiki's repo
+func (repo *Repository) UploadWikiFiles(doer *User, message string, files []string) (err error) {
+	if len(files) == 0 {
+		return nil
+	}
+
+	uploads, err := GetUploadsByUUIDs(files)
+	if err != nil {
+		return fmt.Errorf("GetUploadsByUUIDs [uuids: %v]: %v", files, err)
+	}
+
+	wikiWorkingPool.CheckIn(com.ToStr(repo.ID))
+	defer wikiWorkingPool.CheckOut(com.ToStr(repo.ID))
+
+	if err = repo.InitWiki(); err != nil {
+		return fmt.Errorf("InitWiki: %v", err)
+	}
+
+	localPath := repo.LocalWikiPath()
+
+	if err = discardLocalWikiChanges(localPath); err != nil {
+		return fmt.Errorf("discardLocalWikiChanges: %v", err)
+	} else if err = repo.updateLocalWiki(); err != nil {
+		return fmt.Errorf("UpdateLocalWiki: %v", err)
+	}
+
+	// Copy uploaded files into repository.
+	for _, upload := range uploads {
+		tmpPath := upload.LocalPath()
+		targetPath := path.Join(localPath, upload.Name)
```

Comment:
This will allow you to write to arbitrary places in the filesystem. Admittedly you can't overwrite anything with your current code but you can certainly inject into hooks.

Reply:
Perhaps this could be solved by using the hash of the file path instead of the actual path, in a way similar to what is currently done with avatars (although for avatars it's a hash of the file contents, which I don't think is the best way here).

Reply:
@guillep2k I think if we look for forbidden pathern this shouldnt be a problem. If we hash files on the wiki wont have a human redable name anymore

allow normal chars a-Z numbers and '-' '_' '.' and add a check for `/../` patterns

Reply:
@6543 but then the users will need to think about how they need to name their files. We should allow them to use any name, as long as it renders a serviceable path from Macaron (i.e. no slashes, no empty string).